api/apiroutes.go

```diff
@@ -743,6 +744,73 @@ func (c *appContext) getDecodedTx(w http.ResponseWriter, r *http.Request) {
 	writeJSON(w, tx, m.GetIndentCtx(r))
 }
 
+// getTxSwapInfo checks if the specified transaction is an atomic swap redemption
+// or contract and retrieves information for completed atomic swaps that are
+// related to the transaction.
+func (c *appContext) getTxSwapInfo(w http.ResponseWriter, r *http.Request) {
+	txid, err := m.GetTxIDCtx(r)
+	if err != nil {
+		http.Error(w, http.StatusText(422), 422)
+		return
+	}
+
+	swapTx := &apitypes.AtomicSwapTx{
+		TxID: txid.String(),
+	}
+
+	done := func(err error) {
+		if err != nil {
+			var errString = fmt.Sprintf("Unable to get atomic swap info for transaction %v: %v", txid, err)
+			apiLog.Errorf(errString)
+			http.Error(w, errString, 422)
+		} else {
+			writeJSON(w, swapTx, m.GetIndentCtx(r))
+		}
+	}
+
+	// First check if the passed txid redeems 1 or more atomic swap contracts.
+	swaps, err := rpcutils.AtomicSwapsRedeemedInTx(c.nodeClient, txid.String(), c.Params)
+	if len(swaps) > 0 || err != nil {
+		swapTx.Type = "Redemption"
+		swapTx.Swaps = swaps
+		done(err)
+		return
+	}
+
+	// Retrieve txids that spend this tx's outputs to determine if any of those
+	// spending txs is a redemption tx, which would imply that this tx is a contract tx.
+	spendingTxIDs, _, _, err := c.DataSource.SpendingTransactions(txid.String())
+	if err != nil {
+		done(err)
+		return
+	}
+
+	// It is theoretically possible for a contract tx to have multiple contract
+	// outputs, so check all txs that spend from this tx's outputs and store all
+	// completed swaps returned.
+	allCompletedSwaps := make([]*apitypes.AtomicSwap, 0)
+	for _, spendingTxID := range spendingTxIDs {
+		swaps, err = rpcutils.AtomicSwapsRedeemedInTx(c.nodeClient, spendingTxID, c.Params)
```

Comment:
You're requesting a `TxRawResult` for every spent output over RPC, even if it's not a p2sh script. Since you're already requesting the raw transaction data for `txid` in `AtomicSwapsRedeemedInTx`, maybe keep a hold of that and check that the output pubkey script is p2sh first.